pex/tools/commands/venv.py

```diff
@@ -0,0 +1,236 @@
+    exclude=(),  # type: Tuple[str, ...]
+    collisions_ok=False,  # type: bool
+):
+    # type: (...) -> None
+    safe_mkdir(dst)
+    link = True
+    for root, dirs, files in os.walk(src, topdown=True, followlinks=False):
+        if src == root:
+            dirs[:] = [d for d in dirs if d not in exclude]
+            files[:] = [f for f in files if f not in exclude]
+
+        for d in dirs:
+            try:
+                os.mkdir(os.path.join(dst, os.path.relpath(os.path.join(root, d), src)))
+            except OSError as e:
+                if e.errno != errno.EEXIST:
+                    raise e
+
+        for f in files:
+            src_entry = os.path.join(root, f)
+            dst_entry = os.path.join(dst, os.path.relpath(src_entry, src))
+            try:
+                if link:
+                    try:
+                        os.link(src_entry, dst_entry)
+                        continue
+                    except OSError as e:
+                        if e.errno != errno.EXDEV:
+                            raise e
+                        link = False
+                shutil.copy(src_entry, dst_entry)
+            except OSError as e:
+                if e.errno == errno.EEXIST:
+                    pex_warnings.warn(
+                        "Failed to overwrite {} with {}: {}".format(dst_entry, src_entry, e)
+                    )
+                    if not collisions_ok:
+                        raise e
+
+
+class Venv(Command):
+    """Creates a venv from the PEX file."""
+
+    def add_arguments(self, parser):
+        # type: (ArgumentParser) -> None
+        parser.add_argument(
+            "venv",
+            nargs=1,
+            metavar="PATH",
+            help="The directory to create the virtual environment in.",
+        )
+        parser.add_argument(
+            "-b",
+            "--bin-path",
+            choices=("prepend", "append"),
+            default=None,
+            help="Add the venv bin dir to the PATH in the __main__.py script.",
+        )
+        parser.add_argument(
+            "-f",
+            "--force",
+            action="store_true",
+            default=False,
+            help="If the venv directory already exists, overwrite it.",
+        )
+        parser.add_argument(
+            "--collisions-ok",
+            action="store_true",
+            default=False,
+            help=(
+                "Don't error if population of the venv encounters distributions in the PEX file "
+                "with colliding files."
+            ),
+        )
+        parser.add_argument(
+            "-p",
+            "--pip",
+            action="store_true",
+            default=False,
+            help="Add pip to the venv.",
+        )
+
+    def run(
+        self,
+        pex,  # type: PEX
+        options,  # type: Namespace
+    ):
+        # type: (...) -> Result
+
+        venv = Virtualenv.create(options.venv[0], interpreter=pex.interpreter, force=options.force)
```

Comment:
Since this actually does create the venv on disk as a sideeffect, it might be worth mentioning as step 0 in here.

Reply:
I'm not sure about "side-effect" except in the technical sense that all IO is that way - the method is "create" indicating creation is the main point. That said - done.